### ISSUE TITLE
Add a pyro.isection() context to specify independence

### ISSUE THREAD

**Issue by fritzo** [uber/pyro#983]
This issue proposes a new syntax to specify independence among parts of a program. The proposed syntax `pyro.isection` is similar to the `#pragma omp parallel section` primitive to declare computational independence in OpenMP ([example](http://jakascorner.com/blog/2016/05/omp-sections.html)).

## Motivation
Consider the following dangerous pair of enumerated `iarange`s.
```py
# Dangerous:
with pyro.iarange("enum_1", 3):
    b = pyro.sample("b", dist.Bernoulli(p).reshape([3]))
with pyro.iarange("enum_2", 4):
    c = pyro.sample("c", dist.Bernoulli(p).reshape([4]))
```
In the presence of `@config_enumerate`, it is incorrect for `c` to depend on `b`. In our example, indeed `b` does not depend on `c`, but Pyro cannot recognize this independence. To inform Pyro of the independence, we could use an `irange` as follows:
```py
# Safe but cumbersome:
for i in pyro.irange("switch", 2):
    if i == 0:
        with pyro.iarange("enum_1", 3):
            b = pyro.sample("b", dist.Bernoulli(p).reshape([3]))
    else:
        with pyro.iarange("enum_2", 4):
            c = pyro.sample("c", dist.Bernoulli(p).reshape([4]))
```
Not that this is similar to OpenMP's `sections { section {} section {} }` construct. To make this more usable, we could construct a simple context manager `pyro.isection()` as syntactic sugar for the `irange` construct above:
```py
# Safe and succinct:
section_foo = pyro.isection("foo")
with section_foo, pyro.iarange("enum_1", 3):
    b = pyro.sample("b", dist.Bernoulli(p).reshape([3]))
with section_foo, pyro.iarange("enum_2", 4):
    c = pyro.sample("c", dist.Bernoulli(p).reshape([4]))
```
This is intended to achieve the same `cond_indep_stack` as irange and should require no changes to inference algorithms: it is really only syntax.

**Comment by fritzo** [comment]
@martinjankowiak This seems pretty palatable to me. If we implement this, I think it's safe to raise errors rather than warn when enumeration is unsafe.

**Comment by martinjankowiak** [comment]
dunno about the name but yeah something like that could be good... i suggested broadly similar things in prehistorical time (#172 )

**Comment by eb8680** [comment]
Should be addressed by #1521
